File: src/python/_2130_Maximum_Twin_Sum_of_a_Linked_List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def pairSum(self, head: Optional[ListNode]) -> int:
        prev = None
        slow = head
        fast = head

        while fast:
            fast = fast.next.next
            temp = slow.next
            slow.next = prev
            prev = slow
            slow = temp

        maxSum = 0

        while slow:
            maxSum = max(maxSum, prev.val + slow.val)
            prev = prev.next
            slow = slow.next

        return maxSum

File: src/_2130_Maximum_Twin_Sum_of_a_Linked_List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int pairSum(ListNode head) {
        ListNode prev = null, slow = head, fast = head;

        while (fast != null) {
            fast = fast.next.next;
            ListNode temp = slow.next;
            slow.next = prev;
            prev = slow;
            slow = temp;
        }

        int maxSum = 0;

        while (slow != null) {
            if (slow.val + prev.val > maxSum)
                maxSum = slow.val + prev.val;
            slow = slow.next;
            prev = prev.next;
        }

        System.gc();

        return maxSum;
    }
}
